homeassistant/components/airpatrol/coordinator.py

```diff
@@ -0,0 +1,94 @@
+"""Data update coordinator for AirPatrol."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+from airpatrol.api import AirPatrolAPI, AirPatrolAuthenticationError, AirPatrolError
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, LOGGER, SCAN_INTERVAL
+
+if TYPE_CHECKING:
+    from . import AirPatrolConfigEntry
+
+
+class AirPatrolDataUpdateCoordinator(DataUpdateCoordinator[list[dict[str, Any]]]):
+    """Class to manage fetching AirPatrol data."""
+
+    config_entry: AirPatrolConfigEntry
+    api: AirPatrolAPI | None = None
+
+    def __init__(self, hass: HomeAssistant, config_entry: AirPatrolConfigEntry) -> None:
+        """Initialize."""
+
+        super().__init__(
+            hass,
+            LOGGER,
+            name=f"{DOMAIN.capitalize()} {config_entry.data['email']}",
+            update_interval=SCAN_INTERVAL,
+            config_entry=config_entry,
+        )
+
+    async def _async_update_data(self) -> list[dict[str, Any]]:
+        """Fetch data from API."""
+        try:
+            client = await self.async_get_api_client()
+            return await client.get_data()
+        except AirPatrolAuthenticationError:
+            try:
+                await self._async_refresh_client()
+                client = await self.async_get_api_client()
+                return await client.get_data()
+            except AirPatrolAuthenticationError as refresh_err:
+                raise ConfigEntryAuthFailed(
+                    "Authentication with AirPatrol failed after token refresh"
+                ) from refresh_err
```

Comment:
This is a bit weird with a try block inside another try block.
Put it in a separate function with a retry parameter or something instead.